src/acanban/user.py

```
@@ -0,0 +1,67 @@
+#
+# Acanban is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with Acanban.  If not, see <https://www.gnu.org/licenses/>.
+
+from http import HTTPStatus
+
+from quart import (Blueprint, ResponseReturnValue, abort,
+                   redirect, render_template, request, url_for)
+from quart_auth import current_user
+from rethinkdb.errors import ReqlNonExistenceError
+from werkzeug.exceptions import BadRequestKeyError
+
+from .auth import User
+
+__all__ = ['blueprint']
+
+blueprint = Blueprint('user', __name__, url_prefix='/u')
+
+
+@blueprint.route('/<username>', methods=['GET'])
+async def view_user_profile(username: str) -> ResponseReturnValue:
+    """Handle the user profile page."""
+    user = User(username)
+    try:
+        await user.name
+    except ReqlNonExistenceError:
+        abort(HTTPStatus.NOT_FOUND)
+    return await render_template('user_profile.html', user=user)
+
+
+@blueprint.route('/<username>/edit', methods=['GET', 'POST'])
+async def edit_user_profile(username: str) -> ResponseReturnValue:
+    """Handle the profile edit page."""
+    user = User(username)
+    try:
+        await user.name
+    except ReqlNonExistenceError:
+        abort(HTTPStatus.NOT_FOUND)
+    if user.key != current_user.key:
+        abort(HTTPStatus.FORBIDDEN)
+    if request.method == 'GET':
+        return await render_template('user_profile_edit.html', user=user)
+    info = await request.form
+    try:
+        name, email = info['name'], info['email']
+    except BadRequestKeyError:
+        abort(HTTPStatus.BAD_REQUEST)
+    if name == '':
+        name = await user.name
+    if email == '':
+        email = await user.email
+    await user.update(name=name, email=email)
```

This looks like a lot of unnecessary I/O.

---

This is to avoid making someone's name empty. Maybe this is better in terms of IO:

```python
    if name != '' and email != '':
        await user.update(name=name, email=email)
    elif email == '':
        await user.update(name=name)
    elif name == '':
        await user.update(email=email)
```

---

The number of branches would escalate quickly when there are more details, though, so you'd like to reconsider that.

---

You can always iterate on `info` (with extra care on unexpected fields).

---

Not all fields need this. E.g. bio can be empty.

---

Then you can make use of a mapping, then `**` it into `user.update`.

---

On third thought we should provide better feedback than implicitly falling back to the previous value, e.g. via an error banner.  Same for valid inputs: a S U C C banner should be there.

---

I think that kind of form validation should be on the frontend side (i.e. if the field is empty, there's a warning/error message displayed), and on backend side we still have to do this to protect the data from manual requests. Also, I don't know how to display banners that users can dismiss without using JS.

---

What I meant was on success you return the same page with a success banner (since it's an update operation); otherwise still the same page with an error banner.  The user needs not to dismiss an error, but rather either abort (close tab) or enter the valid input.  That's the general idea for the general case.

In certain occasions (e.g. this one) you can use `pattern` or `requires` to let the browser perform some primitive checking, but that does not prevent one from hand crafting malicious requests and mess up the DB, so we still need to do something in the backend.  Throwing errors instead of silently swallowing them is just my personal preference, for a reason that raising exception is the less cumbersome control flow.

---

I kinda expect to be redirected to the view page rather than the edit page if success.

---

That sounds good too.

---

Do you still need the banner?

---

Probably not, but you need to test the pages yourself to see if the UX is OK.